compiler/rustc_middle/src/ty/print/pretty.rs

```diff
@@ -647,68 +647,97 @@ pub trait PrettyPrinter<'tcx>:
-                            ty::PredicateKind::Projection(projection_predicate) => {
-                                let Some(future_trait) = self.tcx().lang_items().future_trait() else { continue };
-                                let future_output_def_id =
-                                    self.tcx().associated_item_def_ids(future_trait)[0];
-
-                                if projection_predicate.projection_ty.item_def_id
-                                    == future_output_def_id
-                                {
-                                    // We don't account for multiple `Future::Output = Ty` contraints.
-                                    is_future = true;
-                                    future_output_ty = Some(projection_predicate.ty);
-                                }
-                            }
                             ty::PredicateKind::Trait(pred) => {
                                 let trait_ref = bound_predicate.rebind(pred.trait_ref);
+
                                 // Don't print +Sized, but rather +?Sized if absent.
                                 if Some(trait_ref.def_id()) == self.tcx().lang_items().sized_trait()
                                 {
                                     is_sized = true;
                                     continue;
                                 }
 
-                                if Some(trait_ref.def_id())
-                                    == self.tcx().lang_items().future_trait()
+                                // Insert an empty projection list if trait_ref is missing
+                                if !traits.iter().any(|(t, _)| t == &trait_ref) {
+                                    traits.push((trait_ref, vec![]));
+                                }
+                            }
+                            ty::PredicateKind::Projection(pred) => {
+                                let proj_ref = bound_predicate.rebind(pred);
+                                let trait_ref = proj_ref.required_poly_trait_ref(self.tcx());
+
+                                // Projection type entry -- the def-id for naming, and the ty.
+                                let proj_ty =
+                                    (proj_ref.projection_def_id(), proj_ref.skip_binder().ty);
+
+                                // Append to trait_ref's projection list if it exists, otherwise
+                                // insert a new one.
+                                if let Some((_, assoc_items)) =
+                                    traits.iter_mut().find(|(t, _)| t == &trait_ref)
                                 {
-                                    is_future = true;
-                                    continue;
+                                    assoc_items.push(proj_ty);
+                                } else {
+                                    traits.push((trait_ref, vec![proj_ty]));
                                 }
```

Comment:
All of this can be replaced by using the `.entry()` API instead.

Reply:
Yeah, that's what I was using before I hit issues with sorting.

Reply:
To make myself a bit more clear, I was using an HashMap here before and when I iterated over the entries during the printing phase, the test outputs were randomized across runs. I tried to fix this by then collecting back into a Vec only after iteration, but I didn't know what to `sort_by` since all of my keys aren't `Ord`.